---
title: "Introduction to Single Cell RNAseq Part 4"
author: "UCD Bioinformatics Core"
output:
    html_document:
      keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

Last Updated: June 20, 2023

# Part 4: Clustering

## Setup

Load the required libraries and read in the saved object from the previous section.

```{r libraries}
library(Seurat)
library(kableExtra)
library(tidyr)
library(dplyr)
library(ggplot2)
library(dplyr)

experiment.aggregate <- readRDS("scRNA_workshop_3.rds")
experiment.aggregate

set.seed(12345)
```


## Select compenents

In the previous section, we looked at the principal components analysis in a number of different ways. Now we need to select the components that will be used in the following steps (dimensionality reduction by UMAP or tSNE and clustering). If we use too few components, we risk leaving out interesting variation that may define cell types. What happens if we use too many? A greater number of PCs will increase the computational time of following steps.

Lets choose the first 50, based on the elbow plot from the last section.

```{r use.pcs}
use.pcs <- 1:50
```

## Cluster

Seurat implements an graph-based clustering approach. Distances between the cells are calculated based on previously identified PCs.

The default method for identifying k-nearest neighbors has been changed in V4 to [annoy](https://github.com/spotify/annoy) ("Approximate Nearest Neighbors Oh Yeah!). This is an approximate nearest-neighbor approach that is widely used for high-dimensional analysis in many fields, including single-cell analysis. Extensive community benchmarking has shown that annoy substantially improves the speed and memory requirements of neighbor discovery, with negligible impact to downstream results.

Seurat prior approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNAseq data. Briefly, Seurat identified clusters of cells by a shared nearest neighbor (SNN) modularity optimization based clustering algorithm. First calculate k-nearest neighbors (KNN) and construct the SNN graph. Then optimize the modularity function to determine clusters. For a full description of the algorithms, see Waltman and van Eck (2013) The European Physical Journal B. You can switch back to using the previous default setting using nn.method="rann".

The FindClusters function implements the neighbor based clustering procedure, and contains a resolution parameter that sets the granularity of the downstream clustering, with increased values leading to a greater number of clusters. This code produces a series of resolutions for us to investigate and choose from.

```{r find_neighbors_help, eval=FALSE}
?FindNeighbors
```

```{r find_neighbors}
experiment.aggregate <- FindNeighbors(experiment.aggregate, reduction = "pca", dims = use.pcs)
experiment.aggregate <- FindClusters(experiment.aggregate,
                                     resolution = seq(0.25, 4, 0.5))
```

Seurat adds the clustering information to the metadata beginning with "RNA_snn_res." followed by the resolution.

```{r cluster_metadata}
head(experiment.aggregate@meta.data) %>%
  kable(caption = 'Cluster identities are added to the meta.data slot.') %>%
  kable_styling("striped")
```

## Explore clustering resolutions

Lets first investigate how many clusters each resolution produces and set it to the smallest resolutions of 0.5 (fewest clusters).

```{r clusters_per_res}
cluster.resolutions <- grep("res", colnames(experiment.aggregate@meta.data), value = TRUE)
sapply(cluster.resolutions, function(x) length(levels(experiment.aggregate@meta.data[,x])))
```

### Visualize clustering

Dimensionality reduction plots can be used to visualize the clustering results. On these plots, we can see how each clustering resolution aligns with patterns in the data revealed by dimensionality reductions.

#### tSNE

```{r tSNE}
# calculate tSNE
experiment.aggregate <- RunTSNE(experiment.aggregate,
                                reduction.use = "pca",
                                dims = use.pcs,
                                do.fast = TRUE)
# tSNE colored by sample identity
DimPlot(experiment.aggregate,
        group.by = "orig.ident",
        reduction = "tsne",
        shuffle = TRUE) +
  scale_color_viridis_d(option = "mako")
# tSNE colored by cluster
lapply(cluster.resolutions, function(res){
  DimPlot(experiment.aggregate,
          group.by = res,
          reduction = "tsne",
          label = TRUE,
          shuffle = TRUE) +
    scale_color_viridis_d(option = "turbo")
})
```

#### UMAP

```{r UMAP}
# calculate UMAP
experiment.aggregate <- RunUMAP(experiment.aggregate,
                                dims = use.pcs)
# UMAP colored by sample identity
DimPlot(experiment.aggregate,
        group.by = "orig.ident",
        reduction = "umap",
        shuffle = TRUE) +
  scale_color_viridis_d(option = "mako")
# UMAP colored by cluster
lapply(cluster.resolutions, function(res){
  DimPlot(experiment.aggregate,
          group.by = res,
          reduction = "umap",
          label = TRUE,
          shuffle = TRUE) +
    scale_color_viridis_d(option = "turbo")
})
```

### Investigate the relationship between cluster identity and sample identity

```{r membership}
lapply(cluster.resolutions, function(res){
  as.data.frame(table(experiment.aggregate@meta.data[,res], experiment.aggregate$orig.ident))  %>%
    pivot_wider(names_from = Var2, values_from = Freq) %>%
    rename("Cluster" = Var1) %>%
    kable(caption = paste("Resolution", res)) %>%
    kable_styling("striped")
})
lapply(cluster.resolutions, function(res){
         tmp = experiment.aggregate@meta.data[,c(res, "orig.ident")]
         colnames(tmp) = c("cluster", "orig.ident")
         ggplot(tmp, aes(x = cluster, fill = orig.ident)) +
           geom_bar() +
           scale_fill_viridis_d(option = "mako") +
           theme_classic()
})
```

### Visualize metadata

```{r meta}
FeaturePlot(experiment.aggregate,
            reduction = "umap",
            features = c("nCount_RNA", "nFeature_RNA", "percent_MT"))
```

### Visualize cell cycle phase

```{r phase}
DimPlot(experiment.aggregate,
        reduction = "umap",
        group.by = "Phase",
        shuffle = TRUE) +
  scale_color_viridis_d()
```

### Visualize expression of genes of interest

```{r feature}
FeaturePlot(experiment.aggregate,
            reduction = "umap",
            features = "KCNMA1")
```

## Select a resolution

For now, let's use resolution 0.75. Over the remainder of this section, we will refine the clustering further.

```{r select}
Idents(experiment.aggregate) <- "RNA_snn_res.0.75"
```


## Visualize cluster tree

Building a phylogenetic tree relating the 'average' cell from each group in default 'Ident' (currently "RNA_snn_res.1.25"). Tree is estimated based on a distance matrix constructed in either gene expression space or PCA space.

```{r tree}
experiment.aggregate <- BuildClusterTree(experiment.aggregate, dims = use.pcs)
PlotClusterTree(experiment.aggregate)
```


## Merge clusters

In many experiments, the clustering resolution does not need to be uniform across all of the cell types present. While for some cell types of interest fine detail may be desirable, for others, simply grouping them into a larger parent cluster is sufficient. Merging cluster is very straightforward.

```{r merge}
experiment.aggregate <- RenameIdents(experiment.aggregate,
                                     '7' = '4',
                                     '19' = '5',
                                     '20' = '5',
                                     '17' = '14',
                                     '15' = '14',
                                     '16' = '11')
experiment.aggregate$res.0.75_merged <- Idents(experiment.aggregate)
table(experiment.aggregate$res.0.75_merged)
DimPlot(experiment.aggregate,
        reduction = "umap",
        group.by = "res.0.75_merged",
        label = TRUE) +
  scale_color_viridis_d(option = "turbo")
VlnPlot(experiment.aggregate,
        group.by = "res.0.75_merged",
        features = "KCNMA1") +
  scale_fill_viridis_d(option = "turbo")
```

## Reorder the clusters

Merging the clusters changed the order in which they appear on a plot. In order to reorder the clusters for plotting purposes take a look at the levels of the identity, then re-level as desired.

```{r relevel}
levels(experiment.aggregate$res.0.75_merged)
# move one cluster to the first position
experiment.aggregate$res.0.75_merged <- relevel(experiment.aggregate$res.0.75_merged, "0")
levels(experiment.aggregate$res.0.75_merged)
# the color assigned to some clusters will change
VlnPlot(experiment.aggregate,
        group.by = "res.0.75_merged",
        features = "percent_MT") +
  scale_fill_viridis_d(option = "turbo")
# re-level entire factor
new.order <- as.character(sort(as.numeric(levels(experiment.aggregate$res.0.75_merged))))
experiment.aggregate$res.0.75_merged <- factor(experiment.aggregate$res.0.75_merged, levels = new.order)
levels(experiment.aggregate$res.0.75_merged)
VlnPlot(experiment.aggregate,
        group.by = "res.0.75_merged",
        features = "nCount_RNA") +
  scale_fill_viridis_d(option = "turbo")
```

## Subcluster

While merging clusters reduces the resolution in some parts of the experiment, sub-clustering has the opposite effect. Let's produce sub-clusters for cluster 0.
```{r subcluster}
experiment.aggregate <- FindSubCluster(experiment.aggregate,
                                       graph.name = "RNA_snn",
                                       cluster = 0,
                                       subcluster.name = "subcluster")
DimPlot(experiment.aggregate,
        reduction = "umap",
        group.by = "subcluster",
        label = TRUE,
        shuffle = TRUE) +
  scale_color_viridis_d(option = "turbo")
```

## Sub-setting experiment by cluster identity

After exploring and refining the cluster resolution, we may have identified some clusters that are composed of cells we aren't interested in. For example, if we have identified a cluster likely composed of contaminants, this cluster can be removed from the analysis. Alternatively, if a group of clusters have been identified as particularly of interest, these can be isolated and re-analyzed.

```{r subset}
# remove cluster 18
Idents(experiment.aggregate) <- experiment.aggregate$subcluster
experiment.tmp <- subset(experiment.aggregate, subcluster != "18")
DimPlot(experiment.tmp,
        reduction = "umap",
        group.by = "subcluster",
        label = TRUE,
        shuffle = TRUE) +
  scale_color_viridis_d(option = "turbo")
# retain clusters 0_0, 0_1, 0_2, 3, 5, and 13
experiment.tmp <- subset(experiment.aggregate, subcluster %in% c("0_0", "0_1", "0_2", "3", "5", "13"))
DimPlot(experiment.tmp,
        reduction = "umap",
        group.by = "subcluster",
        label = TRUE,
        shuffle = TRUE) +
  scale_color_viridis_d(option = "turbo")
rm(experiment.tmp)
```

## Identify marker genes

Seurat provides several functions that can help you find markers that define clusters via differential expression:

* `FindMarkers` identifies markers for a cluster relative to all other clusters

* `FindAllMarkers` performs the find markers operation for all clusters

* `FindAllMarkersNode` defines all markers that split a node from the cluster tree

### FindMarkers

```{r FindMarkers}
markers <- FindMarkers(experiment.aggregate,
                       group.by = "subcluster",
                       ident.1 = "0_2",
                       ident.2 = c("0_0", "0_1"))
length(which(markers$p_val_adj < 0.05)) # how many are significant?
head(markers) %>%
  kable() %>%
  kable_styling("striped")
```

The "pct.1" and "pct.2" columns record the proportion of cells with normalized expression above 0 in ident.1 and ident.2, respectively. The "p_val" is the raw p-value associated with the differential expression test, while the BH-adjusted value is found in "p_val_adj". Finally, "avg_logFC" is the average log fold change difference between the two groups.

Marker genes identified this way can be visualized in violin plots, feature plots, and heat maps.
```{r view.markers}
view.markers <- c(rownames(markers[markers$avg_log2FC > 0,])[1],
                  rownames(markers[markers$avg_log2FC < 0,])[1])
lapply(view.markers, function(marker){
  VlnPlot(experiment.aggregate,
          group.by = "subcluster",
          features = marker) +
    scale_fill_viridis_d(option = "turbo")
})
FeaturePlot(experiment.aggregate,
            features = view.markers,
            ncol = 2)
DoHeatmap(experiment.aggregate,
          group.by = "subcluster",
          features = view.markers,
          group.colors = viridis::turbo(length(unique(experiment.aggregate$subcluster))))
```


### FindAllMarkers

FindAllMarkers can be used to automate this process across all genes.

```{r FindAllMarkers}
markers <- FindAllMarkers(experiment.aggregate,
                          only.pos = TRUE,
                          min.pct = 0.25,
                          thresh.use = 0.25)
tapply(markers$p_val_adj, markers$cluster, function(x){
  length(x < 0.05)
})
head(markers) %>%
  kable() %>%
  kable_styling("striped")
view.markers <- tapply(markers$gene, markers$cluster, function(x){head(x,1)})
# violin plots
lapply(view.markers, function(marker){
  VlnPlot(experiment.aggregate,
          group.by = "subcluster",
          features = marker) +
    scale_fill_viridis_d(option = "turbo")
})
# feature plots
lapply(view.markers, function(marker){
  FeaturePlot(experiment.aggregate,
              features = marker)
})
# heat map
DoHeatmap(experiment.aggregate,
          group.by = "subcluster",
          features = view.markers,
          group.colors = viridis::turbo(length(unique(experiment.aggregate$subcluster))))
```

#### Calculate mean marker expression within clusters

You may want to get an idea of the mean expression of markers in a cluster or group of clusters. The percent expressing is provided by FindMarkers and FindAllMarkers, along with the average log fold change, but not the expression value itself. The function below calculates a mean for the supplied marker in the named cluster(s) and all other groups. Please note that this function accesses the active identity.

```{r means}
# ensure active identity is set to desired clustering resolution
Idents(experiment.aggregate) <- experiment.aggregate$subcluster
# define function
getGeneClusterMeans <- function(feature, idents){
  x = GetAssayData(experiment.aggregate)[feature,]
  m = tapply(x, Idents(experiment.aggregate) %in% idents, mean)
  names(m) = c("mean.out.of.idents", "mean.in.idents")
  return(m[c(2,1)])
}
# calculate means for a single marker
getGeneClusterMeans("SMOC2", c("1"))
# add means to marker table (example using subset)
markers.small <- markers[view.markers,]
means <- matrix(mapply(getGeneClusterMeans, view.markers, markers.small$cluster), ncol = 2, byrow = TRUE)
colnames(means) <- c("mean.in.cluster", "mean.out.of.cluster")
rownames(means) <- view.markers
markers.small <- cbind(markers.small, means)
markers.small[,c("cluster", "mean.in.cluster", "mean.out.of.cluster", "avg_log2FC", "p_val_adj")] %>%
  kable() %>%
  kable_styling("striped")
```

## Advanced visualizations

Researchers may use the tree, markers, domain knowledge, and goals to identify useful clusters. This may mean adjusting PCA to use, choosing a new resolution, merging clusters together, sub-clustering, sub-setting, etc. You may also want to use automated cell type identification at this point, which will be discussed in the next section.

### Address overplotting
Single cell and single nucleus experiments may include so many cells that dimensionality reduction plots sometimes suffer from overplotting, where individual points are difficult to see. The following code addresses this by adjusting the size and opacity of the points.
```{r alpha}
alpha.use <- 0.4
p <- DimPlot(experiment.aggregate,
             group.by="subcluster",
             pt.size=0.5,
             reduction = "umap",
             shuffle = TRUE) +
  scale_color_viridis_d(option = "turbo")
p$layers[[1]]$mapping$alpha <- alpha.use
p + scale_alpha_continuous(range = alpha.use, guide = FALSE)
```


### Highlight a subset

```{r highlight}
DimPlot(experiment.aggregate,
        group.by = "subcluster",
        cells.highlight = CellsByIdentities(experiment.aggregate, idents = c("0_1", "8", "1")),
        cols.highlight = c(viridis::viridis(3))) +
  ggtitle("Selected sub-clusters")
```

### Split dimensionality reduction plots

```{r split}
DimPlot(experiment.aggregate,
        group.by = "subcluster",
        split.by = "orig.ident") +
  scale_color_viridis_d(option = "turbo")
```

### Plot a subset of cells

Note that the object itself is unchanged by the subsetting operation.
```{r plot.subset}
DimPlot(experiment.aggregate,
        group.by = "subcluster",
        reduction = "umap",
        cells = Cells(experiment.aggregate)[experiment.aggregate$orig.ident %in% "A001-C-007"]) +
  scale_color_viridis_d(option = "turbo") +
  ggtitle("A00-C-007 subcluster")
```

## Save the Seurat object and download the next Rmd file
```{r saveRDS}
# set the finalcluster to subcluster
experiment.aggregate$finalcluster <- experiment.aggregate$subcluster
# save object
saveRDS(experiment.aggregate, file="scRNA_workshop_4.rds")
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2023-June-Single-Cell-RNA-Seq-Analysis/main/data_analysis/scRNA_Workshop-PART5.Rmd", "scRNA_Workshop-PART5.Rmd")
```

## Session Information
```{r session_info}
sessionInfo()
```
